python/nav/eventengine/plugins/upgrade.py

```diff
@@ -0,0 +1,53 @@
+#
+# Copyright (C) 2022 Sikt
+#
+# This file is part of Network Administration Visualized (NAV).
+#
+# NAV is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License version 3 as published by
+# the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+# more details.  You should have received a copy of the GNU General Public
+# License along with NAV. If not, see <http://www.gnu.org/licenses/>.
+#
+"""software upgrade handler plugin"""
+
+from nav.eventengine.plugin import EventHandler
+from nav.eventengine.alerts import AlertGenerator
+
+
+class UpgradeHandler(EventHandler):
+    """Accepts deviceNotice events"""
+
+    handled_types = ('deviceNotice',)
+
+    def handle(self):
+
+        event = self.event
+
+        if event.state != event.STATE_STATELESS:
+            self._logger.info('Ignoring stateful deviceNotice event')
+        else:
+            self._post_alert(event)
+
+        event.delete()
+
+    def _post_alert(self, event):
+        alert = AlertGenerator(event)
+        if alert.alert_type in (
+            "deviceHwUpgrade",
+            "deviceSwUpgrade",
+            "deviceFwUpgrade",
+        ):
+            alert.history_vars = {
+                "old_version": alert["old_version"],
+                "new_version": event.get_subject()
+                .entity_set.filter(software_revision__isnull=False)
+                .first()
+                .software_revision,
```

Comment:
The piece of code (i.e. ipdevpoll) that posted the original event also added the `old_version` value to the event. _I would much rather prefer_ that this same piece of code also added the `new_version` variable, since it already knows it and doesn't have to look it up in the database.

In fact, if the event engine is stopped for some reason, then upon restarting it, it could suddenly be processing two or more upgrade events for the same device, and then both alert entries would set the `new_version` to the currently stored version value, which would only be true for the latest of those two alerts.

I think this plugin should stick to just copy the `new_version` and `old_version` variables as they are, and then you can go back to fixing the ipdevpoll code in a new PR...